signac/core/collection_api.py

```diff
@@ -0,0 +1,491 @@
+        if isinstance(key, _SyncedDict.VALID_KEY_TYPES):
+            key = str(key)
+            if '.' in key:
+                from ..errors import InvalidKeyError
+                raise InvalidKeyError(
+                    "keys may not contain dots ('.'): {}".format(key))
+            else:
+                return key
+        else:
+            from ..errors import KeyTypeError
+            raise KeyTypeError(
+                "keys must be str, int, bool or None, not {}".format(type(key).__name__))
+
+    def __delitem__(self, item):
+        self.load()
+        with self._safe_sync():
+            del self._data[item]
+            self.sync()
+
+    def __setitem__(self, key, value):
+        self.load()
+        with self._safe_sync():
+            with self._suspend_sync():
+                self._data[self._validate_key(key)] = self.from_base(data=value, parent=self)
+            self.sync()
+
+    def __getitem__(self, key):
+        self.load()
+        return self._data[key]
+
+    def __iter__(self):
+        self.load()
+        return iter(self._data)
+
+    def __call__(self):
+        self.load()
+        return self.to_base()
+
+    def __eq__(self, other):
+        if isinstance(other, type(self)):
+            return self() == other()
+        else:
+            return self() == other
+
+    def __len__(self):
+        self.load()
+        return len(self._data)
+
+    def __repr__(self):
+        return repr(self())
+
+    def __str__(self):
+        return str(self())
+
+    def keys(self):
+        self.load()
+        return self._data.keys()
+
+    def values(self):
+        self.load()
+        return self.to_base().values()
+
+    def items(self):
+        self.load()
+        return self.to_base().items()
+
+    def get(self, key, default=None):
+        self.load()
+        return self._data.get(key, default)
+
+    def pop(self, key, default=None):
+        self.load()
+        with self._safe_sync():
+            ret = self._data.pop(key, default)
+            self.sync()
+        return ret
+
+    def popitem(self, key, default=None):
+        self.load()
+        with self._safe_sync():
+            ret = self._data.pop(key, default)
+            self.sync()
+        return ret
+
+    def clear(self):
+        self.load()
+        with self._safe_sync():
+            self._data = {}
+            self.sync()
+
+    def update(self, mapping):
+        self.load()
+        with self._safe_sync():
+            with self._suspend_sync():
+                for key, value in mapping.items():
+                    self[key] = self.from_base(data=value, parent=self)
+            self.sync()
+
+    def setdefault(self, key, default=None):
+        self.load()
+        with self._safe_sync():
+            with self._suspend_sync():
+                ret = self._data.setdefault(key, self.from_base(data=default, parent=self))
+            self.sync()
+        return ret
+
+
+class SyncedAttrDict(_SyncedDict):
```

Comment:
Feel free to merge the classes at this point.

Reply:
I don't think we can merge these. `SyncAttrDict` tries to call SyncedDict's function (`__setattr__`, `__getattr__`, `__delattr__`) and catch the error and the call item function.